### PR TITLE
Enhance DiagnosticDescriptorCreationAnalyzer to support analysis of h…

### PR DESCRIPTION
**Pull request by mavasani**
…elper methods for creating diagnostic descriptors

This ensures that the analyzers in Microsoft.CodeAnalysis.NetAnalyzers which use a helper method for creating descriptors get all the RS diagnostics for correct descriptor usage: RS1017 (DiagnosticIdMustBeAConstant), RS1019 (UseUniqueDiagnosticId) RS1018 (DiagnosticIdMustBeInSpecifiedFormat) and RS1020 (UseCategoriesFromSpecifiedRange)

### CODE REVIEW

**Comment by mavasani** [comment]
Ping @genlu @sharwell @dotpaul for reviews.

File: src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzerTests.cs [hunk 1]
```diff
@@ -291,6 +291,65 @@ public override void Initialize(AnalysisContext context)
...
+            return ImmutableArray.Create(descriptor);
+        }
+    }
+
+    public override void Initialize(AnalysisContext context)
+    {
+    }
+}" + CSharpDiagnosticDescriptorCreationHelper,
+                GetCSharpRS1017ExpectedDiagnostic(14, 43, "descriptor"),
+                GetCSharpRS1019ExpectedDiagnostic(38, 43, "DuplicateDiagnosticId", "MyAnalyzer"));
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzerTests.cs]: 💡 Consider using markup syntax so it's easier to see the full span where these are getting reported.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs [hunk 2]
```diff
@@ -276,28 +269,57 @@ public override void Initialize(AnalysisContext context)
             });
         }
 
-        private static void AnalyzeTitle(OperationAnalysisContext operationAnalysisContext, IObjectCreationOperation objectCreation)
+        private static bool TryGetDescriptorCreateMethodAndArguments(
+            IFieldInitializerOperation fieldInitializer,
+            INamedTypeSymbol diagnosticDescriptorType,
+            out IMethodSymbol creationMethod,
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]:
```suggestion
            [NotNullWhen(true)] out IMethodSymbol? creationMethod,
```

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs [hunk 3]
```diff
@@ -207,32 +207,25 @@ public override void Initialize(AnalysisContext context)
                 var idToAnalyzerMap = new ConcurrentDictionary<string, ConcurrentDictionary<string, ConcurrentBag<Location>>>();
                 compilationContext.RegisterOperationAction(operationAnalysisContext =>
                 {
-                    if (!(((IFieldInitializerOperation)operationAnalysisContext.Operation).Value is IObjectCreationOperation objectCreation))
+                    var fieldInitializer = (IFieldInitializerOperation)operationAnalysisContext.Operation;
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: 💭 Does this work for properties with this form? In theory the backing field initialization could be covered by one of these operations, but I'm not sure if it actually is.

```csharp
static DiagnosticDescriptor Rule { get; } = new DiagnosticDescriptor(...);
```

File: src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzerTests.cs (hunk 1 above)
Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzerTests.cs]: I have kept it consistent with the existing tests in this file. We can clean up the entire file in a follow-up PR.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 2 above)
Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: This file is not yet nullable enabled.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 3 above)
Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Nope, that wasn't supported earlier either. I can file a separate feature request to support that case.

**Comment by mavasani** [comment]
Thanks!

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 2 above)
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: I thought the whole repository was?

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Oh yes, sorry. Switching between Roslyn and RoslynAnalyzers repo confused me. Let me fix this in a follow-up PR.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 3 above)
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Can you file this issue? I believe it's a common pattern for some third-party analyzers.

**Pull request merged**